src/node/mini_miner.cpp

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <node/mini_miner.h>
+
+#include <consensus/amount.h>
+#include <policy/feerate.h>
+#include <primitives/transaction.h>
+#include <timedata.h>
+#include <util/check.h>
+#include <util/moneystr.h>
+
+#include <algorithm>
+#include <numeric>
+#include <utility>
+
+namespace node {
+
+MiniMiner::MiniMiner(const CTxMemPool& mempool, const std::vector<COutPoint>& outpoints)
+{
+    LOCK(mempool.cs);
+    // Find which outpoints to calculate bump fees for.
+    // Anything that's spent by the mempool is to-be-replaced
+    // Anything otherwise unavailable just has a bump fee of 0
+    for (const auto& outpoint : outpoints) {
+        if (const auto ptx{mempool.GetConflictTx(outpoint)}) {
+            // The conflicting transaction is to-be-replaced
+            to_be_replaced.insert(ptx->GetHash());
+        }
```

Comment:
> To check my understanding:

> outpoints here refers to a set of unconfirmed UTXOs that our wallet would like to spend, meaning they are tx outputs. 

They are not necessarily unconfirmed UTXOs, just some specified outpoints/prevouts. They may refer to confirmed or unconfirmed UTXOs. They may refer to outputs that the node doesn't think exist.

> GetConflictTx checks if the outpoint exists in mapNextTx, which means it is the prevout (or input) into another tx in the mempool. If there is another transaction in the mempool spending one of these outputs, we put the tx which is spending the outpointt into to_be_replaced.

correct, `GetConflictTx` gives you another mempool transaction that spends the same tx.

> This is because we plan to construct a transaction which would then replace the conflicting tx, so we want to ignore doing anything with the conflicting txs for now?

We want to ensure we provide a bump fee for this UTXO, but ensure we calculate the bump fees *exluding the to-be-replaced transaction(s)*. Once they're replaced, they won't be there to bump their ancestors. So if we're replacing the child of a CPFP (e.g. to increase the bump even more), we want the correct fee to bump that parent *without* the original child there.

Reply:
Thanks, this explanation really helps!

> They are not necessarily unconfirmed UTXOs, just some specified outpoints/prevouts. They may refer to confirmed or unconfirmed UTXOs. They may refer to outputs that the node doesn't think exist.

Is it correct to say these `outpoints` are intended to be spent in a new transaction, as in the wallet has a set of UTXOs that it wants to spend in `TxB` where `TxB` can be a replacement of an already existing `TxA` or it can be an entirely new tx? From there, these `outpoints` can be a mix of confirmed/unconfirmed, but it's assumed at least one outpoint is unconfirmed

Reply:
Another thought I had while looking at this: we check if an `outpoint` has any conflicting spends in the mempool before we check if the outpoint itself is in the mempool. It seems we are inferring that the `outpoint` is in the mempool, otherwise it would not be present in `mapNextTx`. Is this the safest way to do this? Seems fine because `mapNextTx` is updated every time a tx is removed from the mempool, but figured I'd ask anyway in case there is a race condition / code path I'm not seeing which could make this unreliable in an edge case